File: AppRemover/src/main/java/com/vs_unusedappremover/common/AsyncResult.java
package com.vs_unusedappremover.common;

public class AsyncResult<T> {

	private final Object waitMonitor = new Object();
	private T result;
	private boolean hasResult;
	
	public void set(T result) {
		synchronized (waitMonitor) {
			this.result = result;
			hasResult = true;
			waitMonitor.notify();
		}
	}
	
	public T get() {
		synchronized (waitMonitor) {
			while (!hasResult) {
				try {
					waitMonitor.wait();
				} catch (InterruptedException e) {
					Thread.currentThread().interrupt();
				}
			}
			return result;
		}
	}
}

File: AppRemover/src/main/java/com/vs_unusedappremover/common/AsyncResult.kt
package com.vs_unusedappremover.common

class AsyncResult<T> {

    private val waitMonitor = java.lang.Object()
    private var result: T? = null
    private var hasResult: Boolean = false

    fun set(result: T) {
        synchronized(waitMonitor) {
            this.result = result
            hasResult = true
            waitMonitor.notify()
        }
    }

    fun get(): T? {
        synchronized(waitMonitor) {
            while (!hasResult) {
                try {
                    waitMonitor.wait()
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                }

            }
            return result
        }
    }
}
